multiuser/api/che-multiuser-api-authentication-commons/src/main/java/org/eclipse/che/multiuser/api/authentication/commons/filter/MultiUserEnvironmentInitializationFilter.java

```diff
@@ -146,21 +157,35 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
   }
 
   /**
-   * Calculates id of the user from given authentication token. This <b>may</b> also imply
-   * verification of token signature or encryption for encrypted/signed tokens (like JWT etc)
+   * Processes the token and creates implementation-specific intermediary type using which the
+   * subclasses can extract different kinds of information like user ID or subject.
+   *
+   * <p>This <b>may</b> also imply verification of token signature or encryption for
+   * encrypted/signed tokens (like JWT etc).
+   *
+   * @param token the token to process
+   * @return a processed token or null if the token could not be processed for valid reasons (like
+   *     expiry or not matching any user).
+   */
+  protected abstract Optional<T> processToken(String token);
+
+  /**
+   * Retrieves the id of the user from given authentication token.
    *
-   * @param token authentication token
+   * @param token the original authentication token string
+   * @param processedToken the processed authentication string
    * @return user id given token belongs to
    */
-  protected abstract String getUserId(String token);
+  protected abstract String getUserId(String token, T processedToken);
```

Comment:
If I am not mistaken, none of implementations uses `token` argument, so it may be reduced to `getUserId(T processedToken)`